src/enclave/Enclave/NonObliviousSortMergeJoin.cpp

```diff
@@ -81,10 +80,27 @@ void non_oblivious_sort_merge_join(uint8_t *join_expr, size_t join_expr_length,
   // so last_foreign_row.get() is guaranteed to not be null.
   FlatbuffersTemporaryRow last_foreign_row;
 
+  // A "dummy" row needed for FullOuter joins so that 
+  // any_primary_row.get() is guaranteed to not be null
+  FlatbuffersTemporaryRow any_primary_row;
+  
   while (r.has_next()) {
     const tuix::Row *current = r.next();
+    
     if (current->is_dummy()) {
-      last_foreign_row.set(current);
+      if (join_type == tuix::JoinType_FullOuter) {
+        if (any_primary_row.get() == nullptr) {
+          // Primary table dummy row must come first
+          any_primary_row.set(current);
+        } else if (last_foreign_row.get() == nullptr) {
+          // Foreign table dummy row must come second
```

Comment:
I would make it more explicit that this condition is enforced by the Scala code to avoid confusion.

Reply:
Ok, I'll add better comments and/or try to refactor